src/tim/vx/graph.cc

```diff
@@ -70,20 +72,54 @@ std::shared_ptr<Tensor> GraphImpl::CreateTensorPlaceHolder() {
 }
 
 bool GraphImpl::Compile() {
-  compiled_ = true;
+  bool status = true;
+
+  std::call_once(setio_once_, [&status, this]() {
+    status = (vsi_nn_SetGraphInputs(this->graph_, this->inputs_.data(), this->inputs_.size()) &&
+              vsi_nn_SetGraphOutputs(this->graph_, this->outputs_.data(), this->outputs_.size()));
+  });
+
+  std::call_once(setup_once_, [&status, this](){
+    status = (VSI_SUCCESS == vsi_nn_SetupGraph(this->graph_, true));
+  });
 
-  vsi_nn_SetGraphInputs(graph_, inputs_.data(), inputs_.size());
-  vsi_nn_SetGraphOutputs(graph_, outputs_.data(), outputs_.size());
+  std::call_once(verify_graph_once_, [&status, this]() {
+    status = (VSI_SUCCESS == vsi_nn_VerifyGraph(this->graph_));
+  });
 
-  return (VSI_SUCCESS == vsi_nn_SetupGraph(graph_, true) &&
-          VSI_SUCCESS == vsi_nn_VerifyGraph(graph_));
+  return status;
+}
+
+bool GraphImpl::CompileAsBinary(void* buf, size_t* size) {
+  bool status = true;
+  std::call_once(setio_once_, [&status, this]() {
+    status = (vsi_nn_SetGraphInputs(this->graph_, this->inputs_.data(), this->inputs_.size()) &&
+              vsi_nn_SetGraphOutputs(this->graph_,this->outputs_.data(), this->outputs_.size()));
+  });
+
+  std::call_once(setup_once_, [&status, this](){
+    status = (VSI_SUCCESS == vsi_nn_SetupGraph(this->graph_, true));
+  });
+
+  return ((status) && (VSI_SUCCESS == vsi_nn_GenerateNBG(graph_, buf, size)));
 }
 
 bool GraphImpl::Run() {
-  if (!compiled_ && !Compile()) {
-    return false;
-  }
-  return (VSI_SUCCESS == vsi_nn_RunGraph(graph_));
+  bool status = true;
+  std::call_once(setio_once_, [&status, this]() {
+    status = (vsi_nn_SetGraphInputs(this->graph_, this->inputs_.data(), this->inputs_.size()) &&
+              vsi_nn_SetGraphOutputs(this->graph_, this->outputs_.data(), this->outputs_.size()));
+  });
+
+  std::call_once(setup_once_, [&status, this](){
```

Comment:
Why Run() duplicates all the calls in Compile() again? What's the point of the Compile() function anymore?

Reply:
will change